File: Implementations/DSALG/bst.py
class Tree:
    def __init__(self, root):
        self.root = root
        self.left = None
        self.right = None
        self.arr = []
        
    def search(self, tree, val):
        if tree:
            if tree.root == val:
                return True

            elif val > tree.root:
                return self.search(tree.right, val)

            elif val < tree.root:
                return self.search(tree.left, val)

        else:
            return False

    def add(self, value):
        if self.root:
            if self.root == value:
                return
            else:
                if value < self.root:
                    if self.left:
                        self.left.add(value)
                    else:
                        self.left = Tree(value)
                else:
                    if self.right:
                        self.right.add(value)
                    else:
                        self.right = Tree(value)
        else:
            self.root = Tree(value)
        
    def iot(self, root):
        if root:
            self.iot(root.left)
            self.arr.append(root.root)
            self.iot(root.right)
        
    def pot(self, root):
        if root:
            self.arr.append(root.root)
            self.pot(root.right)
            self.pot(root.left)

    def pot_1(self, root):
        if root:
            self.pot_1(root.left)
            self.pot_1(root.right)
            self.arr.append(root.root)
        
    def show(self):
        arr = self.arr
        self.arr = []
        return arr


if __name__ == '__main__':
    t = Tree(5)
    t.add(2)
    t.add(1)
    t.add(7)
    t.add(6)
    t.add(8)
    t.iot(t)
    print(t.show())
    t.pot(t)
    print(t.show())
    t.pot_1(t)
    print(t.show())
    print(t.search(t, 1))

File: js_basics/bst.js
class Node{
    constructor(num){
        this.val = num;
        this.left = null;
        this.right = null;
    }
}


class BST{
    constructor(num){
        this.root = num;
        this.left = null;
        this.right = null;
        this.arr = [];
    }

    add = function(num){
        var head = this.root;
        if(head == null){
            this.root = new BST(num);
        }else{
            if(num >= this.root){
                if(this.right != null){
                    this.right.add(num);
                }else{
                    this.right = new BST(num);
                }
            }else{
                if(this.left != null){
                    this.left.add(num);
                }else{
                    this.left = new BST(num);
                }
            }
        }
    }

    iot = function(tree){
        if(tree != null){
            this.iot(tree.left);
            console.log(tree.root);
            this.iot(tree.right);
        }
    }

    pot = function(tree){
        if(tree != null){
            console.log(tree.root);
            this.pot(tree.left);
            this.pot(tree.right);
        }
    }

    post = function(tree){
        if(tree != null){
            this.post(tree.left);
            this.post(tree.right);
            console.log(tree.root);
        }
    }

    lot = function(tree){
        var queue = [tree];
        var arr = [];
        var n;
        while(queue.length != 0){
            n = queue.length;
            var temp = [];
            while(n > 0){
                var curNode = queue.shift();
                temp.push(curNode.root);
                if(curNode.left != null){
                    queue.push(curNode.left);
                }
                if(curNode.right != null){
                    queue.push(curNode.right);
                }

                n -= 1;
            }

            arr.push(temp);
        }

        return arr;
    }

    height = function(tree){
        if(tree != null){
            return 1+Math.max(this.height(tree.left), this.height(tree.right));
        }

        return 0;
    }
}


var bst = new BST(5);

bst.add(2);
bst.add(3);
bst.add(1);
bst.add(7);
bst.add(6);
bst.add(8);

// bst.iot(bst);
// bst.pot(bst);
// bst.post(bst);

console.log(bst.lot(bst));

console.log(bst.height(bst));